src/Lib.P2P/SecureCommunication/Secio1Stream.cs

```diff
@@ -0,0 +1,270 @@
+
+            var encryptedData = await ReadPacketBytesAsync(length - _outHmac.GetMacSize(), cancel).ConfigureAwait(false);
+            var signature = await ReadPacketBytesAsync(_outHmac.GetMacSize(), cancel).ConfigureAwait(false);
+
+            var hmac = _outHmac;
+            var mac = new byte[hmac.GetMacSize()];
+            hmac.Reset();
+            hmac.BlockUpdate(encryptedData, 0, encryptedData.Length);
+            hmac.DoFinal(mac, 0);
+            if (!signature.SequenceEqual(mac))
+                throw new InvalidDataException("HMac error");
+
+            // Decrypt the data in-place.
+            _decrypt.ProcessBytes(encryptedData, 0, encryptedData.Length, encryptedData, 0);
+            return encryptedData;
+        }
+
+        private async Task<byte[]> ReadPacketBytesAsync(int count, CancellationToken cancel)
+        {
+            var buffer = new byte[count];
+            await _stream.ReadExactAsync(buffer, 0, count, cancel).ConfigureAwait(false);
+            return buffer;
+        }
+
+        /// <inheritdoc />
+        public override void Flush()
+        {
+#pragma warning disable VSTHRD002 
+            FlushAsync().GetAwaiter().GetResult();
+#pragma warning restore VSTHRD002 
+        }
+
+        /// <inheritdoc />
+        public override async Task FlushAsync(CancellationToken cancel)
+        {
+            if (_outStream.Length == 0)
+                return;
+
+            var data = _outStream.ToArray(); // plain text
+            _encrypt.ProcessBytes(data, 0, data.Length, data, 0);
+
+            var hmac = _inHmac;
+            var mac = new byte[hmac.GetMacSize()];
+            hmac.Reset();
+            hmac.BlockUpdate(data, 0, data.Length);
+            hmac.DoFinal(mac, 0);
+
+            var length = data.Length + mac.Length;
+            _stream.WriteByte((byte) (length >> 24));
+            _stream.WriteByte((byte) (length >> 16));
+            _stream.WriteByte((byte) (length >> 8));
+            _stream.WriteByte((byte) length);
+            await _stream.WriteAsync(data, 0, data.Length, cancel);
+            await _stream.WriteAsync(mac, 0, mac.Length, cancel);
+            await _stream.FlushAsync(cancel).ConfigureAwait(false);
+
+            _outStream.SetLength(0);
+        }
+
+        /// <inheritdoc />
+        public override void Write(byte[] buffer, int offset, int count) { _outStream.Write(buffer, offset, count); }
+
+        /// <inheritdoc />
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return _outStream.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+
+        /// <inheritdoc />
+        public override void WriteByte(byte value) { _outStream.WriteByte(value); }
+
+        /// <inheritdoc />
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing) _stream.Dispose();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)